internal/abci/abci.go

```diff
@@ -835,6 +868,15 @@ func (a *AbciApp) FinalizeBlock(ctx context.Context, req *abciTypes.RequestFinal
 		result.Result.Log += logs.logs
 	}
 
+	// Persist app hash and height to the disk for recovery purposes.
+	lc := &lastCommitInfo{
+		Height:  req.Height,
+		AppHash: newAppHash[:],
+	}
+	if err = lc.saveAs(a.lastCommitInfoFileName); err != nil {
+		return nil, fmt.Errorf("failed to save last commit info: %w", err)
```

Comment:
I'm undecided if this should halt chain.  If nothing else errors, the node is generally OK.  If there is a recovery scenario that would need this file and it's either missing or for the wrong height, it will fail then.
If this ave fails because of a permissions issue (for example), that could be resolved by the operator without even stopping the node.  Thoughts?